Add documentation for Service InternalTrafficPolicy (#27088)
kubernetes/website: content/en/docs/concepts/services-networking/service-traffic-policy.md

---
reviewers:
- maplain
title: Service Internal Traffic Policy
content_type: concept

> weight: 10

Author: updated

---

<!-- overview -->

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

_Service Internal Traffic Policy_ enables cluster internal traffic routing to node
local endpoints only. This approach tries to keep traffic closer to where it
originated from. This can help to reduce costs and improve performance.

<!-- body -->

## Motivation

Kubernetes Service is usually backed up by endpoints scattered around the
whole cluster.
_Service Internal Traffic Policy_ provides a mechanism to help keep cluster
internal traffic within the same node it originated from. The "internal" traffic

> here refers to traffic originated from Pods in the current cluster.

Reviewer: We don't usually have a "motivation" or justification section in our concept docs. 

I would suggest adding the second sentence – 

> "_Service Internal Traffic Policy_ provides a mechanism to help keep cluster internal traffic within the same node it originated from. The "internal" traffic here refers to traffic originated from Pods in the current cluster." 

to the overview section above, and I would delete this section.

Author: updated

## Using Service Internal Traffic Policy

You can enable Internal Traffic Policy for a Service by setting the

> `spec.internalTrafficPolicy` to `Local`. This tells kube-proxy to only use node

Reviewer: Can you provide a code sample/sample yaml in this section? Define the entire service, and then include the `spec.internalTrafficPolicy`.

Author: updated

local endpoints for cluster internal traffic. Importantly, this Service will
appear to have zero endpoint for any Pod accessing this Service on a node, when
this Service doesn't have any local endpoint on the node, but does have endpoints
on other nodes.

## How it Works

kube-proxy filters the endpoints it routes to based on the
`spec.internalTrafficPolicy` setting. When it's "Local", only node local

> endpoints are considered.

Reviewer: What happens when it's not set to `Local`?

Author: updated

---
reviewers:
- maplain
title: Service Internal Traffic Policy
content_type: concept
---

<!-- overview -->

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

> _Service Internal Traffic Policy_ enables cluster internal traffic routing to node
> local endpoints only. This approach provides a mechanism to help keep cluster

Reviewer 2: Non-blocking nit, I found the first sentence confusing as to the benefit of ITP (I think mainly with local endpoint**s** being plural). I don't love my suggestion as it conflicts with the following sentence, but something similar to:

```suggestion
_Service Internal Traffic Policy_ enables restricted Service internal traffic to only the node it originated from. This approach provides a mechanism to help keep cluster
```

Blending the first two sentences together also might work!

Author: updated. let me know your thoughts :) 
@jimangel

Reviewer 2: Reads a lot better IMO thanks!

"internal" traffic here refers to traffic originated from Pods in the current
cluster. This can help to reduce costs and improve performance.

<!-- body -->

## Using Service Internal Traffic Policy

You can enable Internal Traffic Policy for a Service by setting the
`spec.internalTrafficPolicy` to "Local". This tells kube-proxy to only use node
local endpoints for cluster internal traffic. Importantly, this Service will

> appear to have zero endpoint for any Pod accessing this Service on a node, when

Reviewer 3: endpoints (plural)

Author: updated

> this Service doesn't have any local endpoint on the node, but does have endpoints

Reviewer 3: This sentence is a little awkward.  How about:

Importantly, for pods on nodes with no endpoints for a given Service, the Service will behave as if it has zero endpoints (for Pods on this node) even if the Services does have endpoints on other nodes.

??

Author: updated.

> endpoints (for Pods on this node) even if the Services does have endpoints on other

Author: updated

on other nodes.

The following example shows what a Service looks like when internalTrafficPolicy
is set to "Local":

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
  internalTrafficPolicy: Local
```

## How it Works

kube-proxy filters the endpoints it routes to based on the
`spec.internalTrafficPolicy` setting. When it's "Local", only node local
endpoints are considered. When it's "Cluster" or missing, all endpoints are
considered.
When the feature gate `ServiceInternalTrafficPolicy` is on,
`spec.internalTrafficPolicy` defaults to "Cluster".

## Constraints

> * Service Internal Traffic Policy are not used when `externalTrafficPolicy` is set

Reviewer 3: "is not used"

Author: updated

>   to "Local" on a Service. It is possible to use both features in the same cluster

Reviewer 3: I don't recall this corner-case.  It feels wrong - we're going to fix this, right?

Author: this is from https://github.com/kubernetes/website/pull/27032#discussion_r595420521 and https://github.com/kubernetes/kubernetes/issues/100313

Reviewer 3: ACK thanks :)
